### PR TITLE
[Enhancement] AutoFitLabel

### LINKED ISSUE #1136: [Enhancement] Autofit Label
## Summary
A label that grows/shrinks the font size to fit entire bounds

For Android we have [this native API](https://developer.android.com/guide/topics/ui/look-and-feel/autosizing-textview)
For IOS we have [this native API](https://developer.apple.com/documentation/uikit/uilabel/1620546-adjustsfontsizetofitwidth)
For UWP we have [this example](https://github.com/baskren/Forms9Patch/blob/58183808e4e320f3e79c76f919ff28c991160354/Forms9Patch/Forms9Patch.UWP/Renderers/LabeLRenderer.cs#L344):

Ive done a PR on Xamarin Forms for this:
https://github.com/xamarin/Xamarin.Forms/pull/7981

> **Important** this will need be a new control, since we need to change `TextView` to `AppCompatTextView`... [More details here](https://github.com/xamarin/Xamarin.Forms/pull/7981#discussion_r362300397)

## API Changes

- (enum) AutoFitTextMode Label.AutoSizeMode{ get; set; } //Bindable Property
- double Label.MaxAutoFitSize { get; set;} //Bindable Property
- double Label.MinAutoFitSize {get; set;} //Bindable Property

## Intended Use Case
Label to fit bounds and works as a responsive text

```xml
<AutoFitLabel AutoSizeMode="FitToWidth"/>

```
image

## Who Will Do The Work?

- [x] I am willing to take this on myself
- [x] Just putting this out there for someone to pick up

### PR DESCRIPTION
**Pull request by felipebaltazar**
### Bugs Fixed ###

- Fixes #1136 

### API Changes ###

Added: 
 
```csharp
public  class AutoFitLabel {
   AutoFitTextMode Label.AutoSizeMode{ get; set; } //Bindable Property
   double Label.MaxAutoFitSize { get; set;} //Bindable Property
   double Label.MinAutoFitSize {get; set;} //Bindable Property
}
```

### Behavioral Changes ###
Ive implemented a label that adjusts font size according of the label`s size

(I will include some images about this feature)

### PR Checklist ###

- [ ] Has a linked Issue, and the Issue has been `approved`
- [ ] Has tests (if omitted, state reason in description)
- [ ] Has samples (if omitted, state reason in description)
- [ ] Rebased on top of main at time of PR
- [ ] Changes adhere to coding standard

- [ ] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/AutoFitLabelRenderer.android.cs [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+			else
+				SetTextColor(lastUpdateColor.ToAndroid());
+		}
+
+		void UpdateFont()
+		{
+			if (Element is null)
+				return;
+
+#pragma warning disable 618 // We will need to update this when .Font goes away
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/AutoFitLabelRenderer.android.cs:343]: Pls add // TODO:

Comment (AndreiMisiukevich) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/AutoFitLabelRenderer.android.cs:343]: And do we actually need to use obsolete API? Can we use new props?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/AutoFitLabelRenderer.android.cs [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+
+				var yaxis = 0.0;
+				for (var line = startLine; line > 0; line--)
+					yaxis += totalLineHeights[line];
+
+				((ISpatialElement)span).Region = Forms.Region.FromLines(lineHeights, labelWidth, startX, endX, yaxis).Inflate(10);
+			}
+		}
+
+		void SetLineBreakMode(Label label)
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/AutoFitLabelRenderer.android.cs:632]: Can we use the Forms methods? For this and other methods below 
https://github.com/xamarin/Xamarin.Forms/blob/5.0.0/Xamarin.Forms.Platform.Android/Extensions/TextViewExtensions.cs

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/AutoFitLabelRenderer.android.cs [hunk 3]
```diff
@@ -0,0 +1,792 @@
...
+#pragma warning restore 618
+
+					var newTextSize = f.ToScaledPixel();
+					SetTextSize(ComplexUnitType.Sp, newTextSize);
+					break;
+			}
+		}
+	}
+
+	static class BackgroundManager
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/AutoFitLabelRenderer.android.cs:721]: Can you move this class to another file?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/AutoFitLabelRenderer.android.cs [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+				var renderer = sender as IVisualElementRenderer;
+				var control = renderer?.View;
+				var element = renderer?.Element;
+				e.NewElement.PropertyChanged += OnElementPropertyChanged;
+
+				if (control != null && element != null)
+					UpdateBackgroundColor(control, element);
+			}
+
+			Performance.Stop(reference);
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/AutoFitLabelRenderer.android.cs:775]: can you remove this and any other call to `Performance` member?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/FormsAppCompatTextView.android.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+namespace Xamarin.CommunityToolkit.Views
+{
+	public class FormsAppCompatTextView : AppCompatTextView
+	{
+		public FormsAppCompatTextView(Context? context)
+			: base(context)
+		{
+		}
+
+		[Obsolete]
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/Android/FormsAppCompatTextView.android.cs:16]: Any reason to implement these Obsolete members?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/iOS/AutoFitLabelRenderer.ios.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+			base.OnElementPropertyChanged(sender, e);
+
+			if (e.PropertyName == AutoFitLabel.AutoFitModeProperty.PropertyName)
+				UpdateAutoFitMode();
+		}
+
+		void UpdateAutoFitMode()
+		{
+#if __MOBILE__
+			if (Element is AutoFitLabel autoFitLabel)
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/AutoFitLabel/iOS/AutoFitLabelRenderer.ios.cs:58]: Can you invert this `if`?

**Review by pictos — COMMENTED** [review]
@felipebaltazar thanks for this PR. I know this is still a draft PR, so I just did comments regarding the current code, to win time.

**Comment by pictos** [comment]
@felipebaltazar are you able to move this PR foward?

**Comment by felipebaltazar** [comment]
> @felipebaltazar are you able to move this PR foward?

Oh, yeah! Sorry for the delayed response...

 I'm a little busy this month... 
So if anyone knows how to proceed, feel free to commit to this PR... 
I will back to work on this feature soon as possible

**Comment by brminnick** [comment]
Thanks! However, we are no longer adding new features to Xamarin Community Toolkit, focusing on the [.NET MAUI Community Toolkit](https://github.com/communitytoolkit/maui).

Please open a [New Feature Discussion](https://github.com/CommunityToolkit/Maui/discussions/new) to implement this feature in the .NET MAUI Community Toolkit.

I've posted more information about the Future Of Xamarin Community Toolkit here: https://devblogs.microsoft.com/xamarin/the-future-of-xamarin-community-toolkit/?WT.mc_id=mobile-0000-bramin

**Pull request closed without merging**